packages/swapper/src/swappers/JupiterSwapper/swapperApi/getTradeQuote.ts

```diff
@@ -0,0 +1,439 @@
+  TradeRate,
+} from '../../../types'
+import { SwapperName, TradeQuoteError } from '../../../types'
+import { getRate, makeSwapErrorRight } from '../../../utils'
+import type { Instruction } from '../models/Instruction'
+import type { QuoteResponse } from '../models/QuoteResponse'
+import { getJupiterQuote, getJupiterSwapInstructions, isSupportedChainId } from '../utils/helpers'
+
+const tradeQuoteMetadata: Map<string, QuoteResponse> = new Map()
+
+export const getTradeRate = async (
+  input: GetTradeRateInput,
+  deps: SwapperDeps,
+): Promise<Result<TradeRate[], SwapErrorRight>> => {
+  const {
+    sellAsset,
+    buyAsset,
+    sellAmountIncludingProtocolFeesCryptoBaseUnit: sellAmount,
+    affiliateBps,
+    receiveAddress,
+    accountNumber,
+    sendAddress,
+    slippageTolerancePercentageDecimal,
+  } = input
+
+  const { assetsById } = deps
+
+  const jupiterUrl = deps.config.REACT_APP_JUPITER_API_URL
+
+  if (!isSupportedChainId(sellAsset.chainId)) {
+    return Err(
+      makeSwapErrorRight({
+        message: `unsupported chainId`,
+        code: TradeQuoteError.UnsupportedChain,
+        details: { chainId: sellAsset.chainId },
+      }),
+    )
+  }
+
+  if (!isSupportedChainId(buyAsset.chainId)) {
+    return Err(
+      makeSwapErrorRight({
+        message: `unsupported chainId`,
+        code: TradeQuoteError.UnsupportedChain,
+        details: { chainId: sellAsset.chainId },
+      }),
+    )
+  }
+
+  if (!sendAddress) {
+    return Err(
+      makeSwapErrorRight({
+        message: `sendAddress is required`,
+        code: TradeQuoteError.UnknownError,
+        details: { chainId: sellAsset.chainId },
+      }),
+    )
+  }
+
+  const maybeQuoteResponse = await getJupiterQuote({
+    apiUrl: jupiterUrl,
+    sourceAsset: sellAsset.assetId === solAssetId ? wrappedSolAssetId : sellAsset.assetId,
+    destinationAsset: buyAsset.assetId === solAssetId ? wrappedSolAssetId : buyAsset.assetId,
+    commissionBps: affiliateBps,
+    amount: sellAmount,
+    slippageBps: convertDecimalPercentageToBasisPoints(
+      slippageTolerancePercentageDecimal ??
+        getDefaultSlippageDecimalPercentageForSwapper(SwapperName.Jupiter),
+    ).toFixed(),
+  })
+
+  if (maybeQuoteResponse.isErr()) {
+    return Err(
+      makeSwapErrorRight({
+        message: 'Quote request failed',
+        code: TradeQuoteError.NoRouteFound,
+      }),
+    )
+  }
+
+  const { data: quoteResponse } = maybeQuoteResponse.unwrap()
+
+  tradeQuoteMetadata.set('rate', quoteResponse)
+
+  const getFeeData = async () => {
+    const { chainNamespace } = fromAssetId(sellAsset.assetId)
+
+    switch (chainNamespace) {
+      case CHAIN_NAMESPACE.Solana: {
+        const sellAdapter = deps.assertGetSolanaChainAdapter(sellAsset.chainId)
+        const getFeeDataInput: GetFeeDataInput<KnownChainIds.SolanaMainnet> = {
+          to: receiveAddress ?? sendAddress,
+          value: sellAmount,
+          chainSpecific: {
+            from: input.sendAddress!,
```

Comment:
This doesn't look correct either as `sendAddress` is always undefined so non-null assertion won't do anything here either
<img width="472" alt="image" src="https://github.com/user-attachments/assets/540b5c22-7a74-484c-a4a0-7fb6a950cf19">

Reply:
Also tackled